### PR TITLE
EOS-24415: Enable Logrotation for components in the Kubernetes environment.

### PR DESCRIPTION
**Pull request by sumedhak27**
# Problem Statement
- Components should be able to rotate/clean their own logs in a decentralized way.

# Design
-  At each component entry-point the components will add a startup script that configures the log rotation logic.
    Input to this script is config url.
-  Now the configured logic will call the log rotation script/job which will receive the log_path as an input and will rotate the logs.
- https://github.com/Seagate/cortx-experiments/blob/ab81cbdec4983a11446c0934931d2a2dec935e70/logrotation/README.md

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: logrotation-k8env/Dockerfile.cron [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM centos:7.9.2009
```
Comment (ujjwalpl) [inline logrotation-k8env/Dockerfile.cron:1]: Move all the K8s specific stuff into "logrotation/K8s". There should be no "K8s" specific keywords anywhere else outside of this folder.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: logrotation-k8env/k8/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+apiVersion: apps/v1
```
Comment (ujjwalpl) [inline logrotation-k8env/k8/deployment.yaml:1]: Please use "k8s". "k8" is incorrect term.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: logrotation-k8env/logrotate/config/startup.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# /usr/bin/python3
```
Comment (ujjwalpl) [inline logrotation-k8env/logrotate/config/startup.py:1]: This file is not related to logrotate config. In fact this uses logrotate config. Hence needs to be moved out of logrotate.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: logrotation-k8env/Dockerfile.conf_file [hunk 4]
```diff
@@ -0,0 +1,17 @@
+FROM centos:7.9.2009
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply logrotation-k8env/Dockerfile.conf_file]: sure, done 👍

File: logrotation-k8env/Dockerfile.cron (hunk 1 above)
Comment (sumedhak27) [reply logrotation-k8env/Dockerfile.cron]: done

File: logrotation-k8env/k8/deployment.yaml (hunk 2 above)
Comment (sumedhak27) [reply logrotation-k8env/k8/deployment.yaml]: sure 👍 , addressed.

File: logrotation-k8env/logrotate/config/startup.py (hunk 3 above)
Comment (sumedhak27) [reply logrotation-k8env/logrotate/config/startup.py]: Merged the startup.py scripts from ./logrotate/config and ./logrotate/cron dir & move it to /startup/logrotate_setup.py.

File: logrotation/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Enable Log Rotation for Components
+
+Log rotation options for components
```
Comment (cdeshmukh) [inline logrotation/README.md:3]: Why not decide on one approach & propose sample solution for decided one. Any one doesnt suffice all, so two solutions? Can we list down such limitations here, like which component needs what ?

Comment (sumedhak27) [reply logrotation/README.md:3]: Components like motr, s3 use shell scripts for log rotation and hare, csm use logrotate config that's why enabling both ways.

File: logrotation/startup/logrotate_setup.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        if "<COMPONENT_LOG_PATH>" in lines[0]:
+            lines[0] = lines[0].replace("<COMPONENT_LOG_PATH>", log_path)
+    with open(rotate_config, "w") as f:
+        f.truncate()
+        f.writelines(lines)
+# Section end.
+
+
+# Section start: Functions for setup using cron job.
+def run_command(command):
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply logrotation/startup/logrotate_setup.py]: done 👍

File: logrotation/startup/logrotate_setup.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    rotate_conf_file = \
+        "/opt/cortx/component/logrotate/config/log_rollover.conf"
+    component_logpath = "component"
+    log_path = os.path.join(local, component_logpath)
+    update_log_dir(rotate_conf_file, log_path)
+    shutil.move(rotate_conf_file, "/etc/logrotate.d/")
+    shutil.move("/etc/cron.daily/logrotate", "/etc/cron.hourly/")
+
+
+def update_log_dir(rotate_config, log_path):
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply logrotation/startup/logrotate_setup.py]: added the comments.

File: logrotation/startup/logrotate_setup.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+if __name__ == "__main__":
+    argParser = argparse.ArgumentParser()
+    argParser.add_argument(
+        '-c', '--config',
+        help="Config url to read configurations values."
+    )
+    argParser.add_argument(
+        '-m', '--method',
+        help="Method of logrotation, `cronjob` or `logrotate`."
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply logrotation/startup/logrotate_setup.py]: removed this file but two endpoint scripts to enable to both logrotate and cronjob.

File: logrotation/startup/entrypoint.sh [hunk 9]
```diff
@@ -0,0 +1,14 @@
+#! /bin/bash
+
+# Method can be either `cronjob` or `logrotate`.
+# NOTE: The config url should be like "yaml://opt/cortx/component/config.yaml" 
+#       kept it as file dir since cortx-utils is not installed.
+/usr/bin/python3 /opt/cortx/component/startup/logrotate_setup.py \
+--config /opt/cortx/component/config.yaml \
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply logrotation/startup/entrypoint.sh]: updated the format 👍

File: logrotation/logrotate/cron/log_rollover.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
+#! /usr/bin/python3
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply logrotation/logrotate/cron/log_rollover.py]: to simulate components using shell scripts for log rotation.

File: logrotation/README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  - linux cron job
+  - logrotate
+
+## Linux Cron Job
+  - At the entry point,
```
Comment (cdeshmukh) [inline logrotation/README.md:8]: Be more elaborate
'At container entry point..

File: logrotation/README.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    - Run a startup script.\
+    Input to the script is config url.\
+    The script will add a crontab entry to periodically schedule the logrotate script.
+    - Start the crond service.
+    - Run the component entrypoint script.
```
Comment (cdeshmukh) [inline logrotation/README.md:13]: how this entrypoint script is different from startup script (line 9) ?

File: logrotation/README.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+## Logrotate
+  - At the entry point,
+    - Run a startup script.\
+    Input to the script is config url.\
+    The script will fetch the local from config and update the log path accordingly in logrotate configuration file
+    & move the config file to the location /etc/logrotate.d  
+    _**NOTE**: Logrotate applies the config Daily by default, can be forced to Hourly by moving the logrotate script from cron.daily to cron.hourly.\
+    If we want to schedule it more frequently then we need to add a cron job which runs logrotate config with -f flag._
+    - start the crond service.
+    - Run the component entrypoint script.
```
Comment (cdeshmukh) [inline logrotation/README.md:29]: same comment as above

File: logrotation/cron/log_rollover.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    for root, dirs, files in os.walk(log_dir):
+        for file in files:
+            if file[-1].isdigit() and int(file[-1]) > max_i:
+                max_i = int(file[-1])
+        print(max_i)
+        # copy truncates the log file.
+        shutil.copyfile(log_file, f"{log_file}.{max_i+1}")
+        with open(log_file, 'w') as logf:
+            logf.truncate()
+        # rollover is not implemented.
```
Comment (cdeshmukh) [inline logrotation/cron/log_rollover.py:41]: comment applies still? as i see, above logic is for rollover only

File: logrotation/logrotate/log_rollover.conf [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<COMPONENT_LOG_PATH>/*.log {
```
Comment (cdeshmukh) [inline logrotation/logrotate/log_rollover.conf:1]: Add Note here to adjust configs as per component needs. same for any other config file in this PR

File: logrotation/test/py-utils/python_requirements.txt [hunk 16]
```diff
@@ -0,0 +1,34 @@
+aiohttp==3.6.1
```
Comment (cdeshmukh) [inline logrotation/test/py-utils/python_requirements.txt:1]: why this separate requirements file now? Do we need something beyond supplied by cortx-prereq pkg supplied by RE? If we need to add new pkg, need strong justification and approval from Ujjwal over mail, then only RE would add in pre-req pkg

File: logrotation/test/k8s/deployment.yaml [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        command: ["/usr/bin/python3"]
+        args: ["/opt/cortx/component/startup/entrypoint_logrotate.py",
+               "--config", "yaml:///opt/cortx/component/services/confstore.yaml"]
+        volumeMounts:
+        # - name: logger-shared-storage
+        #   mountPath: "/var/log/cortx/"
+        - name: logger-local-storage
+          mountPath: "/var/log/cortx/"
+      nodeSelector:
+        disktype: hdd
```
Comment (cdeshmukh) [inline logrotation/test/k8s/deployment.yaml:36]: why we need this disktype to be mentioned, we need to be agnostic of physical media, from where ever local or shared storage is available for logging

File: logrotation/test/image/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+# Installing required rpm and python pkg
+RUN yum install -y epel-release
+RUN yum install -y gcc rpm-build python36 python36-pip \
+    python36-devel python36-setuptools openssl-devel libffi-devel \
+    python36-dbus crontabs logrotate
+
+COPY ./test/py-utils/ /tmp/
+
+RUN pip3 install -r /tmp/python_requirements.txt
+RUN pip3 install -r /tmp/python_requirements.ext.txt
```
Comment (cdeshmukh) [inline logrotation/test/image/Dockerfile:12]: cortx-prereq pkg wont suffice ?

File: logrotation/startup/entrypoint_logrotate.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    )
+
+    # Step 0: Parse & Validate config url
+    args = argParser.parse_args()
+    if not args.config:
+        usage()
+        exit(1)
+
+    # Step 1: get local log dir from the cofig url.
+    local = get_local(args.config)
```
Comment (cdeshmukh) [inline logrotation/startup/entrypoint_logrotate.py:74]: why local? It can be any path, mapping to a local or shared mount point (dir) ?

File: logrotation/startup/entrypoint_logrotate.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+        lines = f.readlines()
+        if "<COMPONENT_LOG_PATH>" in lines[0]:
+            lines[0] = lines[0].replace("<COMPONENT_LOG_PATH>", log_path)
+    with open(rotate_config, "w") as f:
+        f.truncate()
+        f.writelines(lines)
+
+
+def start_crond_service():
+    cmd = "/usr/sbin/crond start"
```
Comment (cdeshmukh) [inline logrotation/startup/entrypoint_logrotate.py:44]: what happens if this crond service stops after sometime, there is no monitoring for this service rt and running in same component container, so separate eyes for it s availability

File: logrotation/startup/entrypoint_cron.py [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+from cortx.utils.conf_store import Conf
+from cortx.utils.process import SimpleProcess
+import os
+
+sys.path.append('/opt/cortx/component/test')
+from logger import start_logging
+
+
+def get_local(config_url):
+    """Get local log path from the config url."""
```
Comment (cdeshmukh) [inline logrotation/startup/entrypoint_cron.py:14]: no local path necessarily here, can be shared / remote mount point too

File: logrotation/services/confstore.yaml [hunk 22]
```diff
@@ -0,0 +1,4 @@
+cortx:
```
Comment (ujjwalpl) [inline logrotation/services/confstore.yaml:1]: This file can move to startup itself as its related to that and going to be consumed there

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: logrotation/startup/entrypoint_logrotate.py (hunk 19 above)
Comment (ujjwalpl) [reply logrotation/startup/entrypoint_logrotate.py:74]: This should be log path and not local.

File: logrotation/startup/entrypoint_cron.py [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+import sys
+from cortx.utils.conf_store import Conf
+from cortx.utils.process import SimpleProcess
+import os
+
+sys.path.append('/opt/cortx/component/test')
+from logger import start_logging
+
+
+def get_local(config_url):
```
Comment (ujjwalpl) [inline logrotation/startup/entrypoint_cron.py:13]: No need to create a function for one line code. 
Two issues
1. Use Conf.get directly in the code in the beginning. 
2. Use Conf.load in the beginning (execute only once).

File: logrotation/startup/entrypoint_cron.py [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+    )
+
+    # Step 0: Parse & Validate config url
+    args = argParser.parse_args()
+    if not args.config:
+        usage()
+        exit(1)
+
+    # Step 1: get local log dir from the cofig url.
+    local = get_local(args.config)
```
Comment (ujjwalpl) [inline logrotation/startup/entrypoint_cron.py:64]: Use Conf.get(directly)

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: logrotation/test/image/Dockerfile (hunk 18 above)
Comment (sumedhak27) [reply logrotation/test/image/Dockerfile:12]: All the things that I have placed under /test/ dir will not be there with components or in actual implementation Chetan, It is for my testing purpose.
All this installation will already there with cortx-all image.

File: logrotation/test/py-utils/python_requirements.txt (hunk 16 above)
Comment (sumedhak27) [reply logrotation/test/py-utils/python_requirements.txt:1]: It is for my testing, if we want I can remove the whole test folder from the PR, but the experiment alone won't work without that.

File: logrotation/logrotate/log_rollover.conf (hunk 15 above)
Comment (sumedhak27) [reply logrotation/logrotate/log_rollover.conf:1]: done 👍

File: logrotation/startup/entrypoint_logrotate.py (hunk 20 above)
Comment (sumedhak27) [reply logrotation/startup/entrypoint_logrotate.py:38]: Yes, this is true Chetan, are not monitoring this in any way.

File: logrotation/startup/entrypoint_cron.py (hunk 24 above)
Comment (sumedhak27) [reply logrotation/startup/entrypoint_cron.py]: done 👍

File: logrotation/services/confstore.yaml (hunk 22 above)
Comment (sumedhak27) [reply logrotation/services/confstore.yaml]: yes moved it. 👍

File: logrotation/startup/entrypoint_cron.py (hunk 21 above)
Comment (sumedhak27) [reply logrotation/startup/entrypoint_cron.py]: removed this function Chetan. 👍

File: logrotation/startup/entrypoint_logrotate.py (hunk 19 above)
Comment (sumedhak27) [reply logrotation/startup/entrypoint_logrotate.py]: addressed 👍

File: logrotation/test/k8s/deployment.yaml (hunk 17 above)
Comment (sumedhak27) [reply logrotation/test/k8s/deployment.yaml]: sorry, it was just a label I used in my testing. removed it. 👍

File: logrotation/cron/log_rollover.py (hunk 14 above)
Comment (sumedhak27) [reply logrotation/cron/log_rollover.py]: removed 👍

File: logrotation/README.md (hunk 12 above)
Comment (sumedhak27) [reply logrotation/README.md]: rephrased it,  please check. I meant to continue the entry point operations that each component has.

**Review by ujjwalpl — APPROVED** [review]

File: logrotation/startup/entrypoint_logrotate.py (hunk 20 above)
Comment (cdeshmukh) [reply logrotation/startup/entrypoint_logrotate.py:38]: Ok lets raise / discuss this as open item in upcoming discussion with @ujjwalpl

File: logrotation/test/image/Dockerfile (hunk 18 above)
Comment (cdeshmukh) [reply logrotation/test/image/Dockerfile:12]: alright. so you didnt test with cortx-all image

**Review by cdeshmukh — APPROVED** [review]
